### PR TITLE
[Merged by Bors] - feat: add `image_restrictPreimage`

### PR DESCRIPTION
**Pull request by BGuillemet**
- add the theorem `image_restrictPreimage`; use it to simplify the proof of `Set.restrictPreimage_isClosedMap`
- add the equivalent for open maps `Set.restrictPreimage_isOpenMap`.
- delete `IsClosedMap.restrictPreimage`, which duplicates `Set.restrictPreimage_isClosedMap`

---

This PR adds the theorem `image_restrictPreimage` and uses it to simplify the proof of `Set.restrictPreimage_isClosedMap`. It also adds the equivalent for open maps `Set.restrictPreimage_isOpenMap`.
The duplicate `IsClosedMap.restrictPreimage` of `Set.restrictPreimage_isClosedMap` has been deleted.

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by grunweg** [comment]
Welcome to mathlib and thank you for your contribution!

Is your pull request ready for review? Then you can label it `awaiting-review`, so reviewers will look at it :-)
(Have you seen the [contribution guidelines](https://leanprover-community.github.io/contribute/index.html) already? They explain this detail under "Lifecycle of a PR".)

I took the liberty to tweak your PR title and description a bit: for instance, the paragraph you wrote below the line could (imho) also go into the commit message.

File: Mathlib/Data/Set/Function.lean
Comment (grunweg) [inline Mathlib/Data/Set/Function.lean:565]: It's often a good idea to avoid explicit arguments in a variable block, *especially* with hypotheses.
Here, `s` is just a type definition - my intuition is not sure if this is a good idea. I'll let a more experienced reviewer comment.

You certainly could do `variable (s) in` just before your theorem. Whether that's better... well, let's wait for advice, I'm not fully sure any more.

**Comment by BGuillemet** [comment]
Thank you very much for your advice!
I hadn't seen the contribution guidelines, so I read it and added the `awaiting-review` label.

File: Mathlib/Data/Set/Function.lean
Comment (grunweg) [inline Mathlib/Data/Set/Function.lean:573]: Somebody said that code review is about asking questions to the code. I have a question:
can the proof of `range_restrictPreimage` be simplified using this lemma?

(Perhaps the answer is "no, it cannot" or "proving it this way is not simpler" - in these case, one shouldn't do it.)

File: Mathlib/Topology/LocalAtTarget.lean
Comment (grunweg) [inline Mathlib/Topology/LocalAtTarget.lean:69]: Here's a Lean trick: you can simply a tactic proof as a proof term using `by`: in plain English,
you can write "by simp [e, ← image_restrictPreimage]" in place of the last place-holder.

Once you've done that, you have no more goals left, and it's good practice to use `exact instead`.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (grunweg) [inline Mathlib/Topology/LocalAtTarget.lean:75]: You can condense this proof down to one line:
- use the trick above to replace the last two lines by `exact ⟨f '' u, H u hu, by rw [← e, image_restrictPreimage]⟩`
- now, you can switch to providing a term proof instead: instead of by, intro and exact you'd write just
`fun t ⟨u, hu, e⟩ ↦ ⟨f '' u, H u hu, by rw [← e, image_restrictPreimage]⟩` (with no `by` before).

(This is a bit of a detour/general comment. You don't have to memorize this; feel free to read on if you find this interesting.)
In general, it's a matter of style if condensing proofs this way is *desirable*. (If you write the shortest code in the world, but nobody can read it any more, that's not helpful for a library like mathlib which changes often.)
In this case, I'd say it is, as there is not much happening mathematically. In other cases, sometimes it still happens to me that I write code one way and a reviewer suggests a different style: developing "good taste" is always subjective and takes time.

**Review by grunweg — COMMENTED** [review]
I took a look: overall, this PR has a clear goal, does that well and is well-written!
I left three small comments about proof style.

File: Mathlib/Data/Set/Function.lean
Comment (BGuillemet) [reply Mathlib/Data/Set/Function.lean:567]: I simplified the proof of `range_restrictPreimage` using `image_restrictPreimage`. The new proof is indeed way simpler!
However, the proof I wrote seems not to be the more effective one, but I didn't find a way to get rid of the extra `rfl` at the end.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (BGuillemet) [reply Mathlib/Topology/LocalAtTarget.lean]: I shortened this proof the same way I shortened the following one.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (BGuillemet) [reply Mathlib/Topology/LocalAtTarget.lean]: I agree that this proof and the previous one don't need to be three lines long, so I condensed them as you suggested (actually, in this case, I don't even find that shortening these proofs makes them less readable).

File: Mathlib/Topology/LocalAtTarget.lean
Comment (grunweg) [reply Mathlib/Topology/LocalAtTarget.lean]: I was on the fence whether only term mode or two lines ("intro bla; exact baz") is better. :-)
If a maintainer has a strong opinion, we'll see!

**Review by grunweg — APPROVED** [review]
This PR looks broadly good to me. There a few remaining very minor questions where I'm on the fence (see open comments, and perhaps whether the trailing `rfl` can be rewritten more nicely) - a reviewer's/maintainer's opinion is welcome here!

File: Mathlib/Topology/LocalAtTarget.lean
Comment (ADedecker) [inline Mathlib/Topology/LocalAtTarget.lean:69]: This works, but is in some sense not the "morally correct" way to do this, since you're abusing the actual definition of `IsClosed` in terms of `IsOpen`. Could you use [isClosed_induced_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Order.html#isClosed_induced_iff) instead? The proof may get a bit longer, but I'd argue that it's a bit cleaner.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (ADedecker) [inline Mathlib/Topology/LocalAtTarget.lean:73]: If you want the proof to look exactly the same as the closed case you can use [isOpen_induced_iff](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Topology/Order.html#isOpen_induced_iff) instead of relying on the fact that it's true by definition. But really you don't have to, do as you wish.

File: Mathlib/Data/Set/Function.lean
Comment (ADedecker) [reply Mathlib/Data/Set/Function.lean:565]: I don't think it's a problem here because it's a very small section (and indeed this already applies to the current code so I don't think we should consider it a problem). 

Just to be extra clear for Benoit, the issue that could arise is that this would make `s` explicit in a bunch of lemmas below, so when reviewing we have to manually check that it doesn't.

File: Mathlib/Data/Set/Function.lean
Comment (ADedecker) [inline Mathlib/Data/Set/Function.lean:572]:
```suggestion
  rw [← (Subtype.coe_injective).image_injective.eq_iff, ← image_comp, MapsTo.restrict_commutes,
    image_comp, Subtype.image_preimage_coe, Subtype.image_preimage_coe, image_preimage_inter]
```
Slight golf ;)

File: Mathlib/Data/Set/Function.lean
Comment (ADedecker) [inline Mathlib/Data/Set/Function.lean:576]:
```suggestion
  simp only [← image_univ, ← image_restrictPreimage, preimage_univ]
```

File: Mathlib/Data/Set/Function.lean
Comment (ADedecker) [reply Mathlib/Data/Set/Function.lean:567]: Fixed in suggestion below.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (BGuillemet) [reply Mathlib/Topology/LocalAtTarget.lean]: I corrected this proof and the next one as you said. It is indeed slightly longer, but I am not sure that I wrote the most efficient proofs.

File: Mathlib/Topology/Constructions.lean
Comment (jcommelin) [inline Mathlib/Topology/Constructions.lean:1092]: Is this lemma now deleted? Could you add it as (deprecated) alias of `Set.restrictPreimage_isClosedMap`? Arguably it shouldn't even be deprecated, because this name allows to use dot notation.

File: Mathlib/Data/Set/Function.lean
Comment (ADedecker) [reply Mathlib/Data/Set/Function.lean:565]: I think this is fine.

File: Mathlib/Topology/Constructions.lean
Comment (ADedecker) [reply Mathlib/Topology/Constructions.lean:1092]: The point is there was a duplicate, but maybe we want to rename the other version of the lemma to use dot notation.

File: Mathlib/Data/Set/Function.lean
Comment (fpvandoorn) [reply Mathlib/Data/Set/Function.lean:565]: `variable (s) in` is definitely cleaner, and we want to move to that style (and never update binder explicitness globally anymore).
However, it is not your responsibility to do this, since it was already using the suboptimal style before.

File: Mathlib/Topology/Constructions.lean
Comment (fpvandoorn) [reply Mathlib/Topology/Constructions.lean:1092]: Please do add a deprecation, something like
```
@[deprecated] -- Since <date>
alias oldName := newName
```
(I personally like this name better than the other one, so maybe you can rename the other one to `IsClosedMap.restrictPreimage`?)

**Review by fpvandoorn — APPROVED** [review]
LGTM apart from my other comments

File: Mathlib/Topology/LocalAtTarget.lean
Comment (PatrickMassot) [inline Mathlib/Topology/LocalAtTarget.lean:69]: Non-terminal `simp` should be avoided because they make proofs hard to maintain. If someone adds a simp lemma in a file imported by this file then the effect of the above line can change, and then the error on the next line will be painful to debug. 

There are various ways to fix that. 

The lawyer style is to merge those last two lines into `simpa [isClosed_induced_iff] using fun u hu e ↦ ⟨f '' u, H u hu, by simp [← e, image_restrictPreimage]⟩`. This is technically legal because you don’t have a non-terminal `simp`. But of course it suffers from the exact same maintainability issue.

The really explicit way is to replace the `simp` line with:
```lean
  suffices ∀ u, IsClosed u → Subtype.val ⁻¹' u = t → 
    ∃ v, IsClosed v ∧ Subtype.val ⁻¹' v = restrictPreimage s f '' t by 
      simpa [isClosed_induced_iff]
```
which is making explicit what your simp is doing.

The intermediate way is to force the list of lemmas used by simp (queried using `simp?`). That would replace your `simp` line with `simp only [isClosed_induced_iff, forall_exists_index, and_imp]`. It is not nice to see but it solves the maintainability issue.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (PatrickMassot) [inline Mathlib/Topology/LocalAtTarget.lean:75]: Same remark about non-terminal simp here.

**Review by PatrickMassot — COMMENTED** [review]
Nice work! I added two stylistic comments.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (BGuillemet) [reply Mathlib/Topology/LocalAtTarget.lean]: Thank you for your comment. I chose to make explicit what the simp is doing.

File: Mathlib/Topology/LocalAtTarget.lean
Comment (BGuillemet) [reply Mathlib/Topology/LocalAtTarget.lean]: I also made the simp explicit here.

File: Mathlib/Topology/Constructions.lean
Comment (BGuillemet) [reply Mathlib/Topology/Constructions.lean:1092]: I added a deprecation for `Set.restrictPreimage_isClosedMap` and `Set.restrictPreimage_isOpenMap`, so the true names are now `IsClosedMap.restrictPreimage` and `IsOpenMap.restrictPreimage` (with the right order for the arguments).

**Comment by fpvandoorn** [comment]
Thanks!

bors merge

**Pull request closed without merging**
